### PR TITLE
Finish chat

### PR DESCRIPTION
**Pull request by jicc96**
This PR add all the necessary endpoints to the end chat flow, also adds and active job to run a pseudo cron (to be perfected) that finishes all chats that are inactive for more than 10 minutes 

Associated card: https://trello.com/c/wnPAt37A , https://trello.com/c/wyKUoO4n

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -576,5 +582,69 @@ You need the access-token header when calling from the widget.
...
+The chat must exist.
+**Headers**:  
+Content-type: application/json  
+You need the auth_token for the site in the headers with the name access-token!
+
+
+### Put to finish chat by support
+**PATH**
+```
+PUT    {app_domain}/api/v1/chats/:chat_id/finish/client
```
Comment (agupirotto) [inline README.md]: The endpoint is incorrect, the correct endpoint is `PUT    {app_domain}/api/v1/chats/:chat_id/finish/support`

File: app/controllers/api/v1/chats_controller.rb [hunk 2]
```diff
@@ -2,22 +2,123 @@
...
+    return render json: { errors: ['The chat is not in progress.'] }, status: :bad_request unless chat.in_progress?
+
+    return render json: { errors: ["You can't update this chat, you are not assigned to it."] }, status: :forbidden unless current_user.validate_chat_assignment(chat.id) && current_user.validate_site_for_chat(chat.site_id)
+
+    last_client_message = chat.messages.where(user_id: nil).pluck(:created_at).max
+
+    return render json: { errors: ['There are no messages in this chat from the user.'] }, status: :forbidden unless last_client_message.present?
+
+    time_elapsed = ((Time.now.utc - last_client_message.to_time.utc) / 60).to_i
+    return render json: { errors: ["The time passed is #{time_elapsed}, you have to wait until 10 minutes passed"] }, status: :forbidden unless time_elapsed >= 10
```
Comment (agupirotto) [inline app/controllers/api/v1/chats_controller.rb]: You should specify that the time passed is in minutes.

File: app/controllers/api/v1/chats_controller.rb [hunk 3]
```diff
@@ -2,22 +2,123 @@
...
+
+    site = Site.find_by(auth_token: auth_token)
+    return render json: { errors: ['The token provided is invalid.'] }, status: :forbidden unless site.present?
+
+    return render json: { errors: ['You must provide the chat_id.'] }, status: :bad_request unless params[:chat_id].present?
+
+    chat = Chat.find_by(id: params[:chat_id])
+    return render json: { errors: ['Chat not found.'] }, status: :bad_request unless chat.present?
+
+    return render json: { errors: ['Chat not found.'] }, status: :bad_request unless chat.site_id == site.id
```
Comment (agupirotto) [inline app/controllers/api/v1/chats_controller.rb]: This error is that this chat is not belong to the specified site

Comment (jicc96) [reply app/controllers/api/v1/chats_controller.rb]: This was made on purpose.

File: README.md [hunk 4]
```diff
@@ -42,7 +42,11 @@ Note: You can get a new chat_id by calling the **Create a new chat** endpoint
 { 
   body: 'string',
   chat_id: int,
+<<<<<<< HEAD
```
Comment (agupirotto) [inline README.md]: ?

File: app/controllers/api/v1/chats_controller.rb [hunk 5]
```diff
@@ -2,22 +2,123 @@
...
+
+    return render json: { errors: ['The chat is not finished yet.'] }, status: :bad_request unless chat.finished?
+
+    chat.rating = rating
+    return render json: { errors: chat.errors.messages }, status: :not_modified unless chat.save
+
+    render json: { message: 'Thanks for your feedback.' }, status: :ok
```
Comment (agupirotto) [inline app/controllers/api/v1/chats_controller.rb]: You should not modify the rating

File: app/controllers/api/v1/chats_controller.rb [hunk 6]
```diff
@@ -2,22 +2,123 @@
...
+    return render json: { errors: ['Chat not found.'] }, status: :bad_request unless chat.site_id == site.id
+
+    send = { system_request: 'chat_continuation', body: 'The user wishes to continue getting your support.', id: nil, user_id: nil, chat_id: chat.id, created_at: nil }
+    MessagesChannel.broadcast_to(chat, send)
+    head :ok
+  end
+
+  def rating
+    auth_token = request.headers['access-token']
+    return render json: { errors: ['You must provide the auth token.'] }, status: :bad_request unless auth_token.present?
```
Comment (agupirotto) [inline app/controllers/api/v1/chats_controller.rb]: you should change auth-token to access-token

File: app/controllers/api/v1/chats_controller.rb (hunk 5 above)
Comment (agupirotto) [reply app/controllers/api/v1/chats_controller.rb]: Sorry, It works

File: spec/controllers/chats_controller_spec.rb [hunk 7]
```diff
@@ -111,4 +113,171 @@
...
+  describe 'PUT /api/v1/chats/:chat_id/finish/client' do
+    context 'successfully' do
+      it 'rates a chat' do
+        chat.finished!
+        @request.headers['access-token'] = site.auth_token
+        put :rating, params: { chat_id: chat.id, rating: 4 }
+        expect(response).to have_http_status(:ok)
+        expect(JSON.parse(response.body).keys).to contain_exactly('message')
+        expect(JSON.parse(response.body).values).to eq(['Thanks for your feedback.'])
+      end
```
Comment (agupirotto) [inline spec/controllers/chats_controller_spec.rb]: You should add this  `expect(Chat.find(chat.id).rating).to eq(4)`

File: README.md [hunk 8]
```diff
@@ -576,5 +578,69 @@ You need the access-token header when calling from the widget.
...
+The rating is a integer value from 1 to 5.
+
+**Headers**:  
+Content-type: application/json  
+You need the access-token header. 
+
+**Request Body**  
+{
+  "rating": number
+}
```
Comment (agupirotto) [inline README.md]: You must document the new endpoint `continue`

File: app/controllers/api/v1/chats_controller.rb (hunk 2 above)
Comment (jicc96) [reply app/controllers/api/v1/chats_controller.rb]: This was change as requested

**Pull request merged**